### ISSUE TITLE
"gather_facts: yes" ignored in combination with gathering = smart and a custom module

### ISSUE THREAD

**Issue by dhellema** [ansible/ansible#17213]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ansible facts
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```
##### CONFIGURATION

In ansible.cfg:
gathering = smart

Environment:
PYTHONPATH=/srv/ansible/2.1.1/lib:/srv/python-libs
PATH=/srv/ansible/2.1.1/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin
##### OS / ENVIRONMENT

RHEL7
##### SUMMARY

Since Ansible v2 I've noticed the setup module is under certain circumstances skipped regardless of whether "gather_facts: yes" is set. I've written a minimal playbook and module to illustrate the issue. It seems related to the "gathering = smart" setting. Setting it to implicit solves the problem. I also noticed the issue is gone when I don't return "ansible_facts" in the module but custom variables (to be accessed via the "register" construct) instead.
##### STEPS TO REPRODUCE

```
$ cat test.yml
- name: play1
  hosts: all
  gather_facts: no
  tasks:
  - test_bug:
- name: play2
  hosts: all
  gather_facts: yes
  tasks:
  - debug: msg='{{ ansible_virtualization_type }}'


$ cat library/test_bug
#!/usr/bin/python

def main():
   module = AnsibleModule(
      argument_spec = dict()
   )

   dictio={ 'test': 'success' }
   module.exit_json(changed=False, ansible_facts=dictio)

from ansible.module_utils.basic import AnsibleModule
if __name__ == '__main__':
    main()


$ ansible-playbook -l hostname -vv test.yml
```

##### EXPECTED RESULTS

The "gather_facts: yes" should instruct Ansible to retrieve the system specific facts.
##### ACTUAL RESULTS

The setup module is never launched and as a result the playbook fails when trying to use the 'ansible_virtualization_type' variable

```
$ ansible-playbook -l hostname -vv test.yml
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: test.yml *************************************************************
2 plays in test.yml

PLAY [play1] *******************************************************************

TASK [test_bug] ****************************************************************
task path: /home/user/git/ansible/test.yml:5
ok: [el2582] => {"ansible_facts": {"test": "success"}, "changed": false}

PLAY [play2] *******************************************************************

TASK [debug] *******************************************************************
task path: /home/user/git/ansible/test.yml:10
fatal: [hostname]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'ansible_virtualization_type' is undefined\n\nThe error appears to have been in '/home/user/git/ansible/test.yml': line 10, column 5, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n  - debug: msg='{{ ansible_virtualization_type }}'\n    ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes.  Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}

NO MORE HOSTS LEFT *************************************************************
 [WARNING]: Could not create retry file 'test.retry'.         [Errno 2] No such file or directory: ''


PLAY RECAP *********************************************************************
el2582                     : ok=1    changed=0    unreachable=0    failed=1

```

**Comment by alikins** [comment]
I can reproduce this on devel

with gathering = smart:

```
[fedora-23:gather_facts_17213 (master %)]$ ansible --version
ansible 2.2.0 (devel eb33ed4219) last updated 2016/09/27 09:58:42 (GMT -400)
  lib/ansible/modules/core: (detached HEAD c03697c81e) last updated 2016/09/27 09:58:42 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 119bc466be) last updated 2016/09/27 09:58:42 (GMT -400)
  config file = /home/adrian/ansible/ansible-gather-facts-bug.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:gather_facts_17213 (master %)]$ cat ansible-with-gather-smart.cfg 
[defaults]
library        = $HOME/ansible/my-modules:$HOME/ansible/lib/modules/
gathering = smart
[fedora-23:gather_facts_17213 (master %)]$ cat test.yml 
- name: play1
  hosts: all
  gather_facts: no
  tasks:
  - test_bug:
- name: play2
  hosts: all
  gather_facts: yes
  tasks:
  - debug: msg='{{ ansible_virtualization_type }}'
[fedora-23:gather_facts_17213 (master %)]$ ANSIBLE_CONFIG=ansible-with-gather-smart.cfg ansible-playbook -vvv -i hosts test.yml 
Using ansible-with-gather-smart.cfg as config file

PLAYBOOK: test.yml *************************************************************
2 plays in test.yml

PLAY [play1] *******************************************************************

TASK [test_bug] ****************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:5
Using module file /home/adrian/ansible/my-modules/test_bug.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208 `" && echo ansible-tmp-1474990240.27-277895409714208="` echo $HOME/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208 `" ) && sleep 0'
<localhost> PUT /tmp/tmpEtBSHI TO /home/adrian/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208/test_bug.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208/ /home/adrian/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208/test_bug.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208/test_bug.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474990240.27-277895409714208/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "ansible_facts": {
        "test": "success"
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {}, 
        "module_name": "test_bug"
    }
}

PLAY [play2] *******************************************************************

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:10
fatal: [localhost]: FAILED! => {
    "failed": true, 
    "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'ansible_virtualization_type' is undefined\n\nThe error appears to have been in '/home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml': line 10, column 5, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n  - debug: msg='{{ ansible_virtualization_type }}'\n    ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes.  Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"
}
    to retry, use: --limit @/home/adrian/src/ansible-bug-repro/gather_facts_17213/test.retry

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1   

```

**Comment by alikins** [comment]
And without gathering = smart, it works:

```
[fedora-23:gather_facts_17213 (master %)]$ ansible --version
ansible 2.2.0 (devel eb33ed4219) last updated 2016/09/27 09:58:42 (GMT -400)
  lib/ansible/modules/core: (detached HEAD c03697c81e) last updated 2016/09/27 09:58:42 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 119bc466be) last updated 2016/09/27 09:58:42 (GMT -400)
  config file = /home/adrian/ansible/ansible-gather-facts-bug.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:gather_facts_17213 (master %)]$ cat ansible-without-gather-smart.cfg 
[defaults]
library        = $HOME/ansible/my-modules:$HOME/ansible/lib/modules/

[fedora-23:gather_facts_17213 (master %)]$ cat test.yml 
- name: play1
  hosts: all
  gather_facts: no
  tasks:
  - test_bug:
- name: play2
  hosts: all
  gather_facts: yes
  tasks:
  - debug: msg='{{ ansible_virtualization_type }}'
[fedora-23:gather_facts_17213 (master %)]$ ANSIBLE_CONFIG=ansible-without-gather-smart.cfg ansible-playbook -vvv -i hosts test.yml 
Using ansible-without-gather-smart.cfg as config file

PLAYBOOK: test.yml *************************************************************
2 plays in test.yml

PLAY [play1] *******************************************************************

TASK [test_bug] ****************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:5
Using module file /home/adrian/ansible/my-modules/test_bug.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015 `" && echo ansible-tmp-1474990352.36-137474286009015="` echo $HOME/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015 `" ) && sleep 0'
<localhost> PUT /tmp/tmpXdGHli TO /home/adrian/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015/test_bug.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015/ /home/adrian/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015/test_bug.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015/test_bug.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474990352.36-137474286009015/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "ansible_facts": {
        "test": "success"
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {}, 
        "module_name": "test_bug"
    }
}

PLAY [play2] *******************************************************************

TASK [setup] *******************************************************************
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992 `" && echo ansible-tmp-1474990352.65-269773189680992="` echo $HOME/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992 `" ) && sleep 0'
<localhost> PUT /tmp/tmpm7u9aS TO /home/adrian/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992/ /home/adrian/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992/setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992/setup.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474990352.65-269773189680992/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:10
ok: [localhost] => {
    "msg": "kvm"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

**Comment by alikins** [comment]
added repro setup to https://github.com/alikins/ansible-bug-repro/tree/master/gather_facts_17213

**Comment by alikins** [comment]
Using 'gathering = explicit' also works

```
[fedora-23:gather_facts_17213 (master %)]$ ansible --version
ansible 2.2.0 (devel eb33ed4219) last updated 2016/09/27 09:58:42 (GMT -400)
  lib/ansible/modules/core: (detached HEAD c03697c81e) last updated 2016/09/27 09:58:42 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 119bc466be) last updated 2016/09/27 09:58:42 (GMT -400)
  config file = /home/adrian/ansible/ansible-gather-facts-bug.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:gather_facts_17213 (master %)]$ cat ansible-with-gather-explicit.cfg 
[defaults]
#library        = $HOME/ansible/my-modules:$HOME/ansible/lib/modules/
gathering = explicit
[fedora-23:gather_facts_17213 (master %)]$ cat test.yml 
- name: play1
  hosts: all
  gather_facts: no
  tasks:
  - test_bug:
- name: play2
  hosts: all
  gather_facts: yes
  tasks:
  - debug: msg='{{ ansible_virtualization_type }}'
[fedora-23:gather_facts_17213 (master %)]$ ANSIBLE_CONFIG=ansible-without-^Cther-explicit.cfg ansible-playbook -vvv -i hosts test.yml
[fedora-23:gather_facts_17213 (master %)]$ ANSIBLE_CONFIG=ansible-with-gather-explicit.cfg ansible-playbook -vvv -i hosts test.yml
Using ansible-with-gather-explicit.cfg as config file

PLAYBOOK: test.yml *************************************************************
2 plays in test.yml

PLAY [play1] *******************************************************************

TASK [test_bug] ****************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:5
Using module file /home/adrian/src/ansible-bug-repro/gather_facts_17213/library/test_bug.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155 `" && echo ansible-tmp-1474991025.5-3831372058155="` echo $HOME/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155 `" ) && sleep 0'
<localhost> PUT /tmp/tmpRsrVQ3 TO /home/adrian/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155/test_bug.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155/ /home/adrian/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155/test_bug.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155/test_bug.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474991025.5-3831372058155/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "ansible_facts": {
        "test": "success"
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {}, 
        "module_name": "test_bug"
    }
}

PLAY [play2] *******************************************************************

TASK [setup] *******************************************************************
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707 `" && echo ansible-tmp-1474991025.79-265579713031707="` echo $HOME/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707 `" ) && sleep 0'
<localhost> PUT /tmp/tmpJAIFxj TO /home/adrian/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707/ /home/adrian/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707/setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707/setup.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474991025.79-265579713031707/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:10
ok: [localhost] => {
    "msg": "kvm"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0  
```

**Comment by alikins** [comment]
The big if (lots of gather logic) statement in https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/play_iterator.py#L315-L317  seems to end up as False in the gathering=smart case.

In particular, the line at https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/play_iterator.py#L317 is False since 'implied' is False.

**Comment by alikins** [comment]
Removing the 'and implied' from https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/play_iterator.py#L317  seems to make the test case work (and doesn't break the gathering=implicit or gathering=explicit cases)

```
[fedora-23:gather_facts_17213 (master %)]$ ./fails.sh 
Using ansible-with-gather-smart.cfg as config file

PLAYBOOK: test.yml *************************************************************
2 plays in test.yml

PLAY [play1] *******************************************************************
 [WARNING]: about to check gathering

 [WARNING]: gather=smart

 [WARNING]: implied=False

 [WARNING]: gather_facts=False

 [WARNING]: host._gathered_facts=False

 [WARNING]: foo=False

 [WARNING]: bar=False

 [WARNING]: in the big if block

 [WARNING]: in the has_task stanza

 [WARNING]: peek=True

 [WARNING]: peek=False


TASK [setup] *******************************************************************
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003 `" && echo ansible-tmp-1474993434.67-136162085792003="` echo $HOME/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003 `" ) && sleep 0'
<localhost> PUT /tmp/tmp2zi1Dv TO /home/adrian/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003/ /home/adrian/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003/setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003/setup.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474993434.67-136162085792003/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [test_bug] ****************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:5
Using module file /home/adrian/src/ansible-bug-repro/gather_facts_17213/library/test_bug.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654 `" && echo ansible-tmp-1474993435.5-109061456739654="` echo $HOME/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654 `" ) && sleep 0'
<localhost> PUT /tmp/tmpynuiWE TO /home/adrian/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654/test_bug.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654/ /home/adrian/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654/test_bug.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654/test_bug.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1474993435.5-109061456739654/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "ansible_facts": {
        "test": "success"
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {}, 
        "module_name": "test_bug"
    }
}

PLAY [play2] *******************************************************************
 [WARNING]: setting localhost to _gather_facts = True

 [WARNING]: implied=True

 [WARNING]: gather_facts=True

 [WARNING]: host._gathered_facts=True

 [WARNING]: in the empty else of the big if block


TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible-bug-repro/gather_facts_17213/test.yml:10
ok: [localhost] => {
    "msg": "kvm"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

**Comment by alikins** [comment]
https://github.com/ansible/ansible/pull/17774 is a proposed fix. Or at least an implementation of the change mentioned in https://github.com/ansible/ansible/issues/17213#issuecomment-249918267

That change appears to be enough to get the sample reproducer to work, but haven't dug in to see what the implications of the change are.

**Comment by bcoca** [comment]
fixed via 7bdeaac18f1418cb67034e30606dd93c7d498929
